Fixed manual_control usage

Morphlng opened the pull request:
# Description

**!! Still Work In Progress**

We are trying to apply some of our new scenarios to macad-gym and manual control is needed to do the basic scene testing. However, the manual_control feature in macad-gym is completely broken, so I did some changes to make it work. Currently, **it's working again, but I still have some questions**.

# Code changes

## 1. MultiCarlaEnv

1. First of all, if manual control is activated, other render options should be turned off, since PyGame does not allow creating multiple windows in one python process. I'm raising an exception if this kind of conflict occurs, but it is possible to just disable camera render for the user. I wonder what is the best practice in your opinion?

2. Secondly, the whole `_on_render` function doesn't make any sense, or at least, it was called in a wrong place. This function used to be called in `manual_control` branch in `_step` method, so it should display the view for user to control the vehicle. However, it renders all the camera frame in `self._cameras`.

    That brings us to another problem, there isn't actually a camera set up for manual_control right now. The camera bind to an actor is the front camera with 84x84 resolution (default scenario), and the `render_x/y_res` argument used in `manual_control` branch is missing in actor_config. 

    I created some new variable to hold the value needed by manual_control, including:

    ```python
        # For manual_control
        self._display = None
        self._control_clock = None
        self._manual_controller = None
        self._manual_control_camera_manager = None
    ```

    The `_on_render` function is not used right now, because I think it is confusing for people who uses the env to figure out what is the return representing. I think all render related code should be moved to a dedicated class (including the camera_render to deal with the problem I [issued](https://github.com/praveen-palanisamy/macad-gym/issues/72#issue-1457526010) earlier).

## 2. KeyboardControl

This class seems to be copied and modified from Carla's manual_control.py script. The `actor_id` is confusing, it is the name of the actor when creating the Controller, but compared to 1, 2, -1 when parsing keys. I assume this code is meant to be cleaned up but somehow forgotten. I also don't know the difference between those three `parse_keys` functions, so I kept the code for now and use only the default control.

https://github.com/praveen-palanisamy/macad-gym/blob/912944cd93c4db2f704f2818d469425fe191f976/src/macad_gym/core/controllers/keyboard_control.py#L82-L89

## 3. CameraManager & HUD

CameraManager holds the camera, but there isn't an appropriate transform for manual_control right now. I've added a SpringArm transform copied from Carla manual_contrl.py.

HUD is not used a lot (or maybe completely not used?) in macad-gym, so the code is basically not working, especially the `tick` function. The `world` parameter is a wrapper in the original Carla script, but we didn't use it. Now I'm passing everything it needed to update the info, but I'm not satisfied with this solution, do you have any suggestion?

# TODO

1. Move all render related function to a class.
2. Reviewing the code, I found that there are a lot of unused codes or even files. Do we still need to keep them?
3. We are trying to integrate [scenario_runner](https://github.com/carla-simulator/scenario_runner) into macad-gym, but the scenario description, resource management are very different. Any advice on this?

praveen-palanisamy commented:
Hi @Morphlng ,
Good to see your updates on the manual_control mode with detailed PR description.

1.1 
> First of all, if manual control is activated, other render options should be turned off... I wonder what is the best practice in your opinion?

I don't think disabling all the other actor views except the manually-controlled actor view is a good idea. The users would want to have all the sensor views available. If your suggestion is to make the manually-controlled actor's view emphasized because the user is directly controlling it, the `render_x_res`, render_y_res` of that actor's camera view can be set to larger values to make them more prominent for the user to use to control manually.

1.2 The `_on_render(...)` method was meant to be a vehicle-level overridable render function (for example to use a custom resolution for the manually controlled vehicles). I agree that this can be consolidated within `CameraManager.render(...)` or `MultiCarlaEnv._on_render()`.

> there isn't actually a camera set up for manual_control right now...
Every actor can have it's `render_x_res` & `render_y_res` configured via the JSON config whether it is manually controlled or on autopilot or controlled via a custom algorithm. 

> used in manual_control branch is missing in actor_config.

Do you mean the `manual_control.py` script in CARLA Python API? there's no `manual_control` branch on this repository.

2. Yes, the keyboard control sample can use updates and improvements. This repo's main focus is on enabling RL and algorithmic control.

I hope that answers some/most of your questions. Let me know when your work is complete on the PR. Also, it'll be great to add sample scripts to demo the changes or unit tests as part of the PR.

Morphlng commented:
Hi, @praveen-palanisamy!
I've read your suggestions and made a couple of changes.

1. Pygame related **data** is now kept in a dedicated `class Render`. This class is designed to be singleton since only one pygame window is available for one process. However, not all render related logic are kept in here, some of them are still scattered in `multi_env.py`, I've added TODO tag to point them out. I'm not sure whether they should be integrated, the code seems clear enough to me.

2. You can enable camera render and manual_control at the same time now. The view will be something like this:
![manual_control_and_camera](https://user-images.githubusercontent.com/50166287/215266962-ec62e660-5b99-4284-be72-1e813027abd5.gif)

    The screen is resized based on `env_config["render_x/y_res]"`, `env_config["x/y_res]` and how many `actor_config["render"]` is True.

    if no actor is manual controlled or no camera needs to render, or none of them is needed. The view will adjust accordingly, see below:

    - Only camera view:
    <img width="128" alt="only_camera_view" src="https://user-images.githubusercontent.com/50166287/215267430-32cec047-4d8f-4b7d-a1a8-57940a5dedc1.png">

    - Only manual control view: 
    <img width="602" alt="only_manual_control_view" src="https://user-images.githubusercontent.com/50166287/215267437-1c47de6d-38b2-486d-a341-779487aef171.png">

3. Added a brand-new scenario `follow_leading_vehicle.py` to demonstrate how to use manual_control and how to customize an environment (with scenario definition). The above gif shows the scenario view.

4. I went through the project again and listed out some unused files:
    - core/utils/map_explore.py
    - core/vehicle_manager.py
    - carla/spawn_control.py
    - carla/Transform.py
    - Some envs aren't runnable (the config is incompatible with latest release) and they some are not exposed in __init__.py, I'll ignore these for now.

    I wonder if it's ok to just remove them? If so, I'll remove them in next commit. The redundant configuration is also something I want to address, but I think it should be handled in another pull request.

Please review the latest commit and tell me if anything still needs to be improved.

praveen-palanisamy commented on src/macad_gym/viz/render.py:
```diff
@@ -1,70 +1,189 @@
+                exit()
+
+
+if __name__ == "__main__":
```
Good test case! Please convert and move this as a test under `macad_gym/tests/`. We don't need a `__main__` in this module.

praveen-palanisamy commented on src/macad_gym/envs/follow_leading_vehicle.py:
```diff
@@ -0,0 +1,139 @@
+import random
+from macad_gym.envs import MultiCarlaEnv
+
+"""
```
Nice example!
It will be great if you can also add a  [Wiki page here](https://github.com/praveen-palanisamy/macad-gym/wiki) to show how one could import/convert existing Carla/scenario_runner scenarios into RL-enabled MACAD-Gym Envs.

praveen-palanisamy commented on src/macad_gym/envs/follow_leading_vehicle.py:
```diff
@@ -0,0 +1,139 @@
+            "reward_function": "corl2017",
+            "scenarios": "FOLLOWLEADING_TOWN1_CAR1",
+            "manual_control": True,     # manual_control and auto_control
+            "auto_control": True,       # can be turned on at the same time
```
To avoid confusion, please update the comment to say: "When True, starts the actor in auto-pilot and allows manual control take-over on pressing Key `p` on the PyGame window if `manual_control` is also True"

praveen-palanisamy commented on src/macad_gym/envs/follow_leading_vehicle.py:
```diff
@@ -0,0 +1,139 @@
+"""
+Author: Morphlng
+Date: 2023-01-28 18:33:26
+LastEditTime: 2023-01-28 18:33:28
+LastEditors: Morphlng
+Description: Example of creating a custom environment, also a demonstration of how to use the manual_control.
+FilePath: \src\macad_gym\envs\follow_leading_vehicle.py
+"""
```
This file comment header style is not consistent within this repository and most of this information is automatically recorded in git and displayed on GitHub. The description is good to keep as it provides a file-level description.
You can keep your name as `__author__: Morphlng` if you would like.

praveen-palanisamy commented on src/macad_gym/carla/multi_env.py:
```diff
@@ -928,6 +957,23 @@ def _reset(self, clean_world=True):
+                        self, actor_config["auto_control"])
+                    self._manual_controller.actor_id = actor_id
+                    
+                    manual_control_hud = HUD(self._render_x_res, self._render_y_res)
```
can `self._hud` be reused instead of creating a new `HUD` object?

praveen-palanisamy reviewed (changes requested):
Thanks for the PR to add manual control usage with an example!
I left a few minor comments to be addressed before merging. 
Please add a Wiki page showing the manual control usage in action using the `follow_leading_vehicle.py` example. You can re-use the descriptions in this PR. This will make it easier for other to use this feature.

Morphlng commented:
Hi, @praveen-palanisamy!
Code/file changing related requests are completed. I can write a wiki page about how to use manual control soon, but the part:

> It will be great if you can also add a [Wiki page here](https://github.com/praveen-palanisamy/macad-gym/wiki) to show how one could import/convert existing Carla/scenario_runner scenarios into RL-enabled MACAD-Gym Envs.

is not ready yet. The `FollowLeadingVehicle` scenario is hand-crafted using the manual_control ability provided by scenario_runner, since Macad-Gym request a specific start and end position for each vehicle in the scene. Our group is working on combining Macad-Gym and Scenario_Runner right now, I may get back to the wiki page once we're done.

praveen-palanisamy reviewed (approved):
Thanks for addressing the comments and making the changes!
LGTM.

praveen-palanisamy commented:
@Morphlng: Thank you again for your contributions!
I look forward to your integration of Scenario Runner with MACAD-Gym and the wiki updates.
